.gitlab-ci.yml

```diff
@@ -0,0 +1,68 @@
+stages:          # List of stages for jobs, and their order of execution
+  - build-docker
+  - test
+
+workflow:
+  rules:
+    # Only run the pipeline if its for a pull request or the default branch
+    - if: '$CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH'
+    - if: '$CI_EXTERNAL_PULL_REQUEST_IID != null || $CI_EXTERNAL_PULL_REQUEST_IID != ""'
+
+docker_build:
+  stage: build-docker
+  image: docker:20.10
+  tags:
+    - gpu
+    - docker
+  variables:
+    # For this to work, $CI_COMMIT_REF_SLUG == $CI_DEFAULT_BRANCH
+    DOCKER_TAG: $CI_REGISTRY_IMAGE:$CI_COMMIT_REF_SLUG
+  script:
+    - DOCKER_HOST=tcp://localhost:2375 docker login -u $CI_REGISTRY_USER -p $CI_JOB_TOKEN $CI_REGISTRY
+    # Pull the docker containers
+    - docker pull $CI_REGISTRY_IMAGE:$CI_DEFAULT_BRANCH || true
+    - docker pull $DOCKER_TAG || true
+    # Cache from default branch and most recent image to get most out of the caching
+    - docker build -t $DOCKER_TAG --cache-from $CI_REGISTRY_IMAGE:$CI_DEFAULT_BRANCH,$DOCKER_TAG --target tm_base_env .
+    - docker push $DOCKER_TAG
+
+
+memory-tests:
+  stage: test
+  image: $CI_REGISTRY_IMAGE:$CI_COMMIT_REF_SLUG
+  needs: ["docker_build"]
+  tags:
+    - gpu
+  rules:
+    - if: '$CI_EXTERNAL_PULL_REQUEST_IID != null || $CI_EXTERNAL_PULL_REQUEST_IID != ""'
+  variables:
+    CMAKE_ARGS: -DCUDA_ARCH=75
+  script:
+    - CMAKE_BUILD_PARALLEL_LEVEL=$(nproc) pip install -e .[dev,test]
+    - make verify memcheck_tests
+  artifacts:
+    name: "$CI_JOB_NAME-$CI_COMMIT_REF_NAME-memtests"
+    paths:
+        - coverage/
+    when: on_success
+    expire_in: 1 week
+
+unit-tests:
+  stage: test
+  image: $CI_REGISTRY_IMAGE:$CI_COMMIT_REF_SLUG
+  needs: ["docker_build"]
+  tags:
+    - gpu
+  rules:
+    - if: '$CI_EXTERNAL_PULL_REQUEST_IID != null || $CI_EXTERNAL_PULL_REQUEST_IID != ""'
+  variables:
+    CMAKE_ARGS: -DCUDA_ARCH=75
+  script:
+    - CMAKE_BUILD_PARALLEL_LEVEL=$(nproc) pip install -e .[dev,test]
+    - make verify tests
```

Comment:
Would it make sense to run `verify` in a separate stage that runs first? I can think of 2 advantages:
- we'd get linting feedback right away (vs. after the docker build)
- wouldn't need to run `verify` in two places (`unit-tests` and `memory-tests`)